imap/cyr_expire.c

```diff
@@ -282,43 +283,114 @@ static int parse_duration(const char *s, int *secondsp)
     int ret = 0;
     /* mboxname needs to be copied since `mboxname_make_parent`
      * runs a strrchr() on it.
      */
-    char *buf = xstrdup(mbentry->name);
+    char *buf = xstrdup(mboxname);
 
     /*
      * Mailboxes inherit /vendo/cmu/cyrus-imapd/{expire, archive, delete},
      * so we need to iterate all the way up to "" (server entry).
      */
     do {
-        buf_free(&attrib);
-        ret = annotatemore_lookup(buf, annot_entry, "", &attrib);
+        buf_reset(annot_value);
+        ret = annotatemore_lookup(buf, annot_entry, "", annot_value);
         if (ret ||              /* error */
-            attrib.s)           /* found an entry */
+            buf_len(annot_value))           /* found an entry */
             break;
     } while (mboxname_make_parent(buf) && iterate);
 
-    if (attrib.s && parse_duration(attrib.s, secondsp))
+    free(buf);
+
+    return buf_len(annot_value) ? 1 : 0;
+}
+
+static int get_duration_annotation(const char *mboxname,
+                                   const char *annot_entry,
+                                   int *secondsp, bool iterate)
+{
+    struct buf attrib = BUF_INITIALIZER;
+    int ret = 0;
+
+    if (get_annotation_value(mboxname, annot_entry, &attrib, iterate) &&
+            parse_duration(buf_cstring(&attrib), secondsp))
         ret = 1;
-    else
-        ret = 0;
 
     buf_free(&attrib);
-    free(buf);
+    return ret;
+}
+
+static int get_time_annotation(const char *mboxname,
+                               const char *annot_entry,
+                               time_t *timep, bool iterate)
+{
+    struct buf attrib = BUF_INITIALIZER;
+    int ret = 0;
+
+    if (get_annotation_value(mboxname, annot_entry, &attrib, iterate)) {
+        const char *end = NULL;
+        bit64 v64 = 0;
+        if (!parsenum(buf_cstring(&attrib), &end, 0, &v64) && !*end) {
+            *timep = v64;
+            ret = 1;
+        }
+    }
 
+    buf_free(&attrib);
+    return ret;
+}
+
+static int noexpire_mailbox(const mbentry_t *mbentry)
+{
+    int ret = 0;
+    mbname_t *mbname = mbname_from_intname(mbentry->name);
+
+    if (mbname_userid(mbname)) {
+        // Cache result for the last seen userid
+        static struct {
+            struct buf userid;
+            int has_noexpire;
+        } last_seen = { BUF_INITIALIZER, -1 };
+
+        if (!strcmp(mbname_userid(mbname), buf_cstring(&last_seen.userid))) {
+            ret = last_seen.has_noexpire;
+            goto done;
+        }
+
+        // Determine user inbox name
+        if (mbname_isdeleted(mbname)) {
+            mbname_t *tmp = mbname_from_userid(mbname_userid(mbname));
+            mbname_free(&mbname);
+            mbname = tmp;
+        }
+        mbname_truncate_boxes(mbname, 0);
+
+        // Lookup annotation
+        time_t until;
+        if (get_time_annotation(mbname_intname(mbname),
+                    IMAP_ANNOT_NS "noexpire_until", &until, false))
+            ret = !until || progtime < until;
```

Comment:
These conditionals should be fine. Quoting [GNU libc documentation](https://www.gnu.org/software/libc/manual/html_node/Time-Types.html):

> On POSIX-conformant systems, time_t is an integer type and its values represent the number of seconds elapsed since the epoch, which is 00:00:00 on January 1, 1970, Coordinated Universal Time. 

Also, the annotation value type in annotate.c is defined be an unsigned integer, so any negative value will be rejected on write.

I'm not sure if there's a need to break up the lines, still?

Reply:
In that case it might be worth putting a comment there, so the next reader doesn't need to wrack their brain wondering whether there's a bad edge case for pre-1970 dates (negative time_t)

Reply:
Fixed with https://github.com/cyrusimap/cyrus-imapd/pull/4533/commits/e9ec21a79394289d7280ed7d657c810b224eaefb#diff-e2b440985f7fec0ccf9584ee3e7606debbfd031652c5e1188af23d7079091659R380-R384